### PR TITLE
qgis3: fix Processing/GRASS

### PR DESCRIPTION
**Pull request by fjperini**
Solution for the problem reported in #436.

### CODE REVIEW

**Comment by nickrobison** [comment]
@fjperini Thanks for working to track down the Grass window issue!

I'm not sure I understand what this PR is doing. Are you disabling Grass support, regardless of whether or not the build flag is enabled? That seems like the wrong solution, is Grass support totally broken, or is it merely that the use might need to dismiss a window on startup?

Also, Head should always point to the current tip of master, not a specific commit, that's the purpose of being able to install the current latest, even if there may be other build issues.

**Comment by fjperini** [comment]
@nickrobison I'm working on the formula. I hope to upload the changes over the course of the night.
Everything is working correctly!

**Comment by nickrobison** [comment]
@fjperini Disabling Grass is not the correct solution. If the Grass plugin is working, we need to keep supporting it, even if it means the user has to dismiss the popup window until 3.2.2 is released.

**Comment by fjperini** [comment]
@nickrobison It is not deactivated. The problem is already solved.

File: Formula/qgis3.rb [hunk 1]
```diff
@@ -374,41 +409,55 @@ def install
       -DWITH_INTERNAL_QWTPOLAR=FALSE
       -DQGIS_MACAPP_BUNDLE=0
       -DQGIS_MACAPP_INSTALL_DEV=FALSE
-      -DWITH_QSCIAPI=TRUE
+      -DWITH_QSCIAPI=FALSE
       -DWITH_STAGED_PLUGINS=TRUE
       -DWITH_CUSTOM_WIDGETS=TRUE
       -DWITH_ASTYLE=FALSE
-      -DUSE_CCACHE=OFF
+      -DWITH_GRASS=FALSE
```
Comment (nickrobison) [inline Formula/qgis3.rb]: Is this the solution to the Grass windowing problem?

Comment (fjperini) [reply Formula/qgis3.rb]: No, that disables GRASS 6, the spoorte we use is WITH_GRASS7.
The problem was in some `def` and `path`.

Comment (fjperini) [reply Formula/qgis3.rb]: You can see that it is used in the same way as in `qgis2`.

https://github.com/OSGeo/homebrew-osgeo4mac/blob/17bf22a08f46f7a17bac9b7f5615c26157283d04/Formula/qgis2.rb#L285

Comment (nickrobison) [reply Formula/qgis3.rb]: Ok, that makes sense. My apologies.

So in the PR, what is the actual fix to the windowing issue?

**Comment by fjperini** [comment]
```
def python_site_packages
    HOMEBREW_PREFIX/"lib/python#{py_ver}/site-packages"
end
```
It seems that this is the cause`libexec/.../site-packages` If used, the window appears,

**Comment by fjperini** [comment]
I'm making some changes because, although they are installed, the verification does not find them.

```
# check for required run-time Python module dependencies
# TODO: add "pyspatialite" when PyPi package supports spatialite 4.x
  xm = []
   %w[psycopg2 matplotlib pyparsing requests future jinja2 pygments].each do |m|
   xm << m unless module_importable? m
end
```
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'psycopg2'
...
```

I think it has to do with

```
# handle custom site-packages for keg-only modules and packages
ENV.append_path "PYTHONPATH", python_site_packages
...
```

**Comment by fjperini** [comment]
@nickrobison I think it's better to eliminate that in `caveats`, since the modules are installed correctly and the program works perfectly.

**Comment by fjperini** [comment]
I have solved the message problem of the modules.
Later I apply the changes.

**Comment by nickrobison** [comment]
@fjperini @luisspuerto I'm having a difficult time following all the comments in the thread, can you please summarize for me what the current state of this ticket is?

Are we still experiencing the Grass window popping up?

Are the python3 modules being loaded correctly?

File: Formula/qgis3.rb [hunk 2]
```diff
@@ -374,41 +409,55 @@ def install
...
+    ENV["PYTHONHOME"] = python_prefix
 
-    # handle custom site-packages for qt keg-only modules and packages
-    ENV.prepend_path "PYTHONPATH", python_site_packages
-    ENV.append_path "PYTHONPATH", python_qt_site_packages
-    # ENV.append_path "PYTHONPATH", libexec/"vendor/lib/python#{py_ver}/site-packages"
-    ENV.prepend_path "PATH", libexec/'vendor/bin/'
-    # ENV.prepend_path "PATH", libexec/"python/bin"
+    # handle custom site-packages for keg-only modules and packages
+    ENV.append_path "PYTHONPATH", python_site_packages
```
Comment (nickrobison) [inline Formula/qgis3.rb]: This seems to entirely defeat the purpose of vendoring our python resources. If we simply add the system site-packages into the path, before adding the vendored directory to the path, then we're relying on whatever packages and versions the users has installed. I don't see how that's better than vendoring the resources ourselves.

File: Formula/qgis3.rb [hunk 3]
```diff
@@ -809,53 +865,33 @@ def brewed_grass7?
...
 
   def brewed_python?
-    Formula["python3"].linked_keg.exist?
+    Formula["python"].linked_keg.exist?
   end
 
   def python_exec
     if brewed_python?
-      Formula["python3"].opt_bin/"python3"
+      Formula["python"].opt_bin/"python3"
```
Comment (nickrobison) [inline Formula/qgis3.rb]: If we're prepending the vendored python to the path, shouldn't we set that as the python executable for the rest of the formula to use?

File: Formula/qgis3.rb (hunk 2 above)
Comment (fjperini) [reply Formula/qgis3.rb]: I solved the pop-up window with
```
# handle custom site-packages for keg-only modules and packages
ENV.append_path "PYTHONPATH", libexec/"python/lib/python#{py_ver}/site-packages"
ENV.append_path "PYTHONPATH", lib/"python#{py_ver}/site-packages"
```
```
# set install's lib/python#{py_ver}/site-packages first, so app will work if unlinked
pypths = %W[
  #{opt_lib}/python#{py_ver}/site-packages
  #{opt_libexec}/python/lib/python#{py_ver}/site-packages
  #{pypth}
]
```

File: Formula/qgis3.rb (hunk 3 above)
Comment (fjperini) [reply Formula/qgis3.rb]: You say use: `libexec/"python/bin/pyton3"`

Comment (nickrobison) [reply Formula/qgis3.rb]: That would seem to be the correct approach. If we're relying on the vendored python packages, we should use the vendored python installation as well.

Comment (fjperini) [reply Formula/qgis3.rb]: Working on minor changes, later I upload everything.
And I will test what happens with the GRASS scripts comparing building with `python 3`.

Comment (nickrobison) [reply Formula/qgis3.rb]: Ok. Thanks!

Comment (nickrobison) [reply Formula/qgis3.rb]: When building GRASS with python3 please make sure we don't remove support for python2 scripts.

Comment (fjperini) [reply Formula/qgis3.rb]: The changes I made in the GRASS formula

```
# ensure python3 is used
bin.env_script_all_files(libexec/"bin", :GRASS_PYTHON => "python3")
```
```
def formula_site_packages(f)
  `python3 -c "import os, sys, site; sp1 = list(sys.path); site.addsitedir('#{Formula[f].opt_lib}/python3.7/site-packages'); print(os.pathsep.join([x for x in sys.path if x not in sp1]))"`.strip
end
```

Comment (nickrobison) [reply Formula/qgis3.rb]: When you're ready, open a PR and we can go through it. Trying to keep these ticket treads from getting too long.

Comment (fjperini) [reply Formula/qgis3.rb]: Ok, let's close this PR.

**Comment by fjperini** [comment]
@nickrobison Changes applied!
In summary, the problem with the window is that you need `libexec/vendor/bin` in PATH or `/usr/local/lib/python3.7/site-packages` in PYTHONPATH. Otherwise it appears. It may be a GRASS problem!
Extremely tested!
I chose to add in PATH.
As for the modules, they are installed and QGIS works correctly.

**Comment by nickrobison** [comment]
Looks good to me. Thanks for all your hard work!

**Pull request merged**